frontend/ui/widget/virtualkeyboard.lua

```diff
@@ -45,16 +47,29 @@ local VirtualKey = InputContainer:new{
 
 function VirtualKey:init()
     if self.keyboard.symbolmode_keys[self.label] ~= nil then
-        self.callback = function () self.keyboard:setLayout("Sym") end
+        self.callback = function () self.keyboard:setLayer("Sym") end
         self.skiptap = true
     elseif self.keyboard.shiftmode_keys[self.label] ~= nil then
-        self.callback = function () self.keyboard:setLayout("Shift") end
+        self.callback = function () self.keyboard:setLayer("Shift") end
         self.skiptap = true
     elseif self.keyboard.utf8mode_keys[self.label] ~= nil then
-        self.callback = function () self.keyboard:setLayout("IM") end
+        self.callback = function ()
+            self.keyboard_layout_dialog = KeyboardLayoutDialog:new{
+                parent = self,
+            }
+            UIManager:show(self.keyboard_layout_dialog)
+        end
+        self.hold_callback = function()
+            self.key_chars = self:genkeyboardLayoutPopup(self)
```

Comment:
I'd personally prefer if these 2 were inverted:
Tap to get the keypopup and quick switching.
Hold to show your new widget only when needed, when I need another language I have not yet checked to be in the keypopup.

Reply:
These key_chars need to be outside hold_callback to they can also be used in swipe_callback.

(I've been trying to hint at swipes for a bit but I guess I wasn't explicit enough. :-P)

Reply:
as of now:
swipe works.
tap brings the `VirtualKeyPopup` where the gobe icon will bring up the full popup
hold brings up the whole popup

does that work?

Reply:
also if there are no active keyboards, swipe does nothing and tap brings up the full popup